File: backend/src/prechecks/concrete/cluster/master-eligible-nodes.precheck.ts
import { ElasticClient } from "../../../clients/elastic.client";
import { BaseClusterPrecheck } from "../../base/base-cluster-precheck";
import { ExecutionMode, PrecheckType } from "../../types/enums";
import { ClusterContext, PrecheckExecutionRequest } from "../../types/interfaces";

export class MasterEligibleNodesPrecheck extends BaseClusterPrecheck {
	constructor() {
		super({
			id: "elasticsearch_master_eligible_nodes_check",
			name: "Minimum number of master-eligible nodes",
			type: PrecheckType.CLUSTER,
			mode: ExecutionMode.CODE,
		});
	}

	protected async runForContext(request: PrecheckExecutionRequest<ClusterContext>): Promise<void> {
		const client = await ElasticClient.buildClient(request.cluster.clusterId);
		const es = client.getClient();

		const nodes = await es.cat.nodes({ format: "json", h: "name,node.role" });
		const totalNodes = nodes.length;
		const masterEligibleNodes = nodes.filter((n) => n["node.role"]?.includes("m"));
		const masterCount = masterEligibleNodes.length;

		this.addLog(request, `Found ${masterCount} master-eligible node(s) out of ${totalNodes} total node(s).`);

		const isSmallCluster = totalNodes < 3;

		if (isSmallCluster) {
			this.addLog(
				request,
				`ℹ️ Small cluster detected (${totalNodes} total node${totalNodes > 1 ? "s" : ""}). ` +
					`For production high availability, it's recommended to have at least 3 master-eligible nodes.`
			);
			return;
		}

		if (masterCount % 2 === 0) {
			this.addLog(
				request,
				`⚠️ Even number (${masterCount}) of master-eligible nodes detected. ` +
					`Consider using an odd number (e.g., ${masterCount + 1}) to avoid split-brain scenarios.`
			);
		}

		const quorum = Math.floor(masterCount / 2) + 1;
		const toleratedFailures = masterCount - quorum;

		this.addLog(
			request,
			`🧮 Master quorum requirement: ${quorum} out of ${masterCount} master-eligible nodes. ` +
				`Cluster can tolerate up to ${toleratedFailures} failure(s) and still elect a master.`
		);

		if (masterCount < 3) {
			this.addLog(
				request,
				`❌ Only ${masterCount} master-eligible node${masterCount > 1 ? "s" : ""} detected. ` +
					`This is below the recommended minimum of 3 for safe master elections.`
			);
			throw new Error(
				`Insufficient master-eligible nodes: found ${masterCount}, need at least 3 for high availability.`
			);
		}
	}
}

File: server/modules/precheck/src/main/java/co/hyperflex/precheck/concrete/cluster/MasterEligibleNodesPrecheck.java
package co.hyperflex.precheck.concrete.cluster;

import co.hyperflex.clients.elastic.ElasticClient;
import co.hyperflex.clients.elastic.dto.nodes.NodeInfo;
import co.hyperflex.clients.elastic.dto.nodes.NodeRole;
import co.hyperflex.precheck.contexts.ClusterContext;
import co.hyperflex.precheck.core.BaseClusterPrecheck;
import co.hyperflex.precheck.core.enums.PrecheckSeverity;
import java.util.Collection;
import java.util.List;
import java.util.function.Predicate;
import org.slf4j.Logger;
import org.springframework.stereotype.Component;

@Component
public class MasterEligibleNodesPrecheck extends BaseClusterPrecheck {

  @Override
  public String getName() {
    return "Minimum number of master-eligible nodes";
  }

  @Override
  public PrecheckSeverity getSeverity() {
    return PrecheckSeverity.WARNING;
  }

  @Override
  public void run(ClusterContext context) {
    ElasticClient client = context.getElasticClient();
    Logger logger = context.getLogger();

    Collection<NodeInfo> nodes = client.getNodesInfo().getNodes().values();

    int totalNodes = nodes.size();

    List<NodeInfo> masterEligibleNodes =
        nodes.stream().filter(isMasterNode())
            .toList();

    int masterCount = masterEligibleNodes.size();

    logger.info("Found {} master-eligible node(s) out of {} total node(s).", masterCount, totalNodes);

    boolean isSmallCluster = totalNodes <= 3;

    if (isSmallCluster) {
      logger.info("Small cluster detected ({} total node{}). For production high availability,"
          + " it's recommended to have at least 3 master-eligible nodes.", totalNodes, totalNodes > 1 ? "s" : "");
      return;
    }

    if (masterCount % 2 == 0) {
      logger.warn("Even number ({}) of master-eligible nodes detected. Consider using an "
          + "odd number (e.g., {}) to avoid split-brain scenarios.", masterCount, masterCount + 1);
    }

    int quorum = (masterCount / 2) + 1;
    int toleratedFailures = masterCount - quorum;

    logger.info("Master quorum requirement: {} out of {} master-eligible nodes."
        + " Cluster can tolerate up to {} failure(s) and still elect a master.", quorum, masterCount, toleratedFailures);

    if (masterCount < 3) {
      logger.error("Only {} master-eligible node{} detected. This is below the recommended minimum of 3 for safe master elections.",
          masterCount, masterCount > 1 ? "s" : "");
      logger.error("Insufficient master-eligible nodes: found {} need at least 3 for high availability.", masterCount);
      throw new RuntimeException();
    }

  }

  private Predicate<NodeInfo> isMasterNode() {
    return n -> n.getRoles() != null && n.getRoles().contains(NodeRole.Master);
  }
}
